### ISSUE TITLE
Music

### ISSUE THREAD

**Issue by tube0013** [sammachin/AlexaPi#20]
Is is possible to get Amazon's Music Services working with the script?  Could you use something like ffmpeg or vlc to stream handle the streams?

**Comment by lennysh** [comment]
It's being worked on by me at least...

**Comment by groetingc** [comment]
Could You please explain me in some words what I need to do that Music-Service is also working for me?

**Comment by tube0013** [comment]
I got approved for Music on my amazon dev account, so if there is any code to test I'd be happy to try.

**Comment by SpiderSpartan** [comment]
Amazon has an [Alexa service implementation guide](https://developer.amazon.com/public/solutions/alexa/alexa-voice-service/docs/reference-implementation-guide), I haven't had a chance to dig into this, but it does reference and walk you through installing VLC on linux.  If I ever do get around to reading through all this and get music working I'll share what I find.

**Comment by dliloch** [comment]
I too got approved for Music on my amazon dev account and would be happy to try. Looking at the implementation guide it appears that you have to create a "Music Event" then I guess you would 
point it to VLC. VLC is installed on my PI but I don't know how to code the linkage between the main.py code and the VLC.. If I ask Alexa to play a song "she" says fine but of course nothing. Before I was approved I would get an unauthorized for this service message...

**Comment by lennysh** [comment]
Music is very tricky as every track has to be requested individually. And you don't need to request the next track until the previous track is done. I've got this working in my .Net version already.

**Comment by dliloch** [comment]
this is the article I was referring to:
https://developer.amazon.com/public/solutions/alexa/alexa-voice-service/rest/audioplayer-events-response#Audio%20Item%20JSON%20Object

**Comment by lennysh** [comment]
The other issue is the progress reporting. It has set parameters on when to do the first report, and how far apart the next reports need to be.

**Comment by lennysh** [comment]
Just played my first song from Amazon Prime Music! 

Now just to get the progress reporting working so the next song will play properly.

**Comment by dliloch** [comment]
Wow...amazing.. It is Python I assume... Can I try it?

**Comment by dliloch** [comment]
Did you see this email from Amazon..

Valued AVS Developer,

We are excited to announce an important new update to the Alexa Voice Service (AVS) that will enable you to enhance the user experience on your Alexa-enabled products. Last year, we launched a developer preview of AVS to introduce you to the benefits of voice-powered experiences. With this update, we are making architectural improvements that include updated APIs and message structures, Amazon Alexa App (iOS and Android) support and the ability to send server-initiated messages.

Here are some of the new features:

Volume Control: Ability to adjust and mute/unmute volume on your product using voice or buttons.
Media Control: Ability to control media playback (e.g. play/pause) on your product using voice or buttons.
Timers and Alarms: Ability to create and manage timers/alarms on your product using voice.
Amazon Alexa App Enhancements: Control volume, media playback (e.g. Amazon Music, Audible), as well as manage timers/alarms on your product using the Amazon Alexa App.
Learn how to integrate Alexa into your product with the Getting Started Guide. If you already have a solution, see the API Migration Guide(v20160207).

Have Questions? We are here to help! Visit us on the AVS Forum to discuss specific questions with one of our experts.

Alexa is always learning, and there is a lot more to come!

- Alexa Voice Service Team

**Comment by lennysh** [comment]
It's the same script but modified. 

I got to iron out a few issues, and then I'll post the code.

**Comment by dliloch** [comment]
Thanks so much..

**Comment by dliloch** [comment]
Lenny,
One thing I noticed with the existing code is it sort of does a spin loop on the button GPIO pin. I wonder if it could wait on the pin to change state..on a pi 2 one of the cores is pegged constantly...ie. Instead of polling use GPIO.wait_for_edge(18,GPIO.RISING)?

**Comment by lennysh** [comment]
I think someone already posted a pull request making this very change.  I still need to look over it myself to see how it performs in comparison. I'm also assuming Sam hasn't had much of a chance to look it over either in order to merge it.

**Comment by lennysh** [comment]
I've merged my changes into my music-patch-1 branch located [here](https://github.com/lennysh/AlexaPi/tree/lennysh-music-patch-1).

There is a boolean at the top to enable debug mode if you have any issues.  Please enable it prior to uploading any script output for troubleshooting.

I'm not going to send a pull request yet because Sam is going to update his script to work with the new API first.  Once that is done, and I get my side updated the same way, then I'll do a pull request unless asked otherwise.

**Comment by dliloch** [comment]
downloaded the code and redid the setup script. Then I tried it and it
said connect OK but I did not hear the voice. I then asked for the time
and it took about a minute to come back. I did not have this issue with the
original script, which I still have and it still works fine (it is in
another directory). I asked ALEXA to play fleetwood mac and it appears to
connect but no music. I set debug=1 and I've attached the  output.
Please let me know if you need any further diagnostics.
Thanks for all your effort here ... this is beyond my coding capabilities
.. but I can contribute by testing...
this is running on a raspy-3 using the wii-fi connection..

-don

On Thu, Mar 24, 2016 at 8:37 AM, Lenny Shirley <notifications@github.com>
wrote:

> I've merged my changes into my music-patch-1 branch located here
> <https://github.com/lennysh/AlexaPi/tree/lennysh-music-patch-1>.
>
> There is a boolean at the top to enable debug mode if you have any issues.
> Please enable it prior to uploading any script output for troubleshooting.
>
> I'm not going to send a pull request yet because Sam is going to update
> his script to work with the new API first. Once that is done, and I get my
> side updated the same way, then I'll do a pull request unless asked
> otherwise.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sammachin/AlexaPi/issues/20#issuecomment-200813290>
>

comparing to the main.py without music...
alexa's response to questions very slow
when I first start it up python main.py .. it connects but I do not get the
audio on the hello.mp3..
I asked to play fleetwood mac and as you can see it appeared to connect but no
sound ..

pi@raspialexa:~/Downloads/AlexaPi-lennysh-music-patch-1 $ python main.pyChecking Internet Connection...
Connection OK
Play_Audio Request for: hello.mp3
Player State: State.Opening
Player State: State.Playing
Player State: State.Ended
Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Notifications#ACRI#DeviceTTSRenderer_07573134-f5e3-4bbe-8ce1-0232a2ec865b","audioContent":"cid:DeviceTTSRenderer_07573134-f5e3-4bbe-8ce1-0232a2ec865b_1050288615"}}]}}
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Player State: State.Ended
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Knowledge#ACRI#KnowledgePrompt.f40c15bf-1aa7-4f7b-a4d6-7d44f0765581","audioContent":"cid:KnowledgePrompt.f40c15bf-1aa7-4f7b-a4d6-7d44f0765581_664830369"}}]}}
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Player State: State.Ended
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Knowledge#ACRI#KnowledgePrompt.487ce0c6-a3aa-400a-a75d-79410c9b2eb0","audioContent":"cid:KnowledgePrompt.487ce0c6-a3aa-400a-a75d-79410c9b2eb0_1863796510"}}]}}
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Player State: State.Ended
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_ac21c594-3b04-4276-a3d8-3326ffdacce9","audioContent":"cid:DeviceTTSRendererV4_ac21c594-3b04-4276-a3d8-3326ffdacce9_728356133"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=203496952&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=8c3885d96034192d43a153f41b32a72149f2ba0ee035e3789daa22e02bca2e2b","offsetInMilliseconds":0,"expiryTime":"2016-03-24T15:06:57+0000","progressReport":{"progressReportDelayInMilliseconds":1000,"progressReportIntervalInMilliseconds":0},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#3d81d0bf-d958-44d7-89c3-07aa3d2a11ff:54"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#3d81d0bf-d958-44d7-89c3-07aa3d2a11ff:54"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#3d81d0bf-d958-44d7-89c3-07aa3d2a11ff:54","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#3d81d0bf-d958-44d7-89c3-07aa3d2a11ff:54
streamUrl: https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=203496952&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=8c3885d96034192d43a153f41b32a72149f2ba0ee035e3789daa22e02bca2e2b
streamId: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#3d81d0bf-d958-44d7-89c3-07aa3d2a11ff:54
playBehavior: REPLACE_PREVIOUS
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Sending Playback Progress Report Request...
Player State: State.Ended
Sending Playback Progress Report Request...
 Play_Audio Request for: https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=203496952&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=8c3885d96034192d43a153f41b32a72149f2ba0ee035e3789daa22e02bca2e2b
Player State: State.Opening
Playback Progress Report was Successful!Playback Progress Report was Successful!

[74419c50] httplive stream: HTTP Live Streaming (d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=203496952&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=8c3885d96034192d43a153f41b32a72149f2ba0ee035e3789daa22e02bca2e2b)
Recording...
Player State: State.Stopped
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[]}}
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_9e9155db-6d58-476d-87e2-75b64cb4e442","audioContent":"cid:DeviceTTSRendererV4_9e9155db-6d58-476d-87e2-75b64cb4e442_74032601"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=258196818&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=6fe8ccce6c53dafa52946f2272ffecea11f4d2635b2fd218d95326739cd96c53","offsetInMilliseconds":0,"expiryTime":"2016-03-24T15:09:36+0000","progressReport":{"progressReportDelayInMilliseconds":1000,"progressReportIntervalInMilliseconds":0},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#36eaf4b1-8df8-4707-89cf-063d2a7435a4:1"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#36eaf4b1-8df8-4707-89cf-063d2a7435a4:1"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#36eaf4b1-8df8-4707-89cf-063d2a7435a4:1","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#36eaf4b1-8df8-4707-89cf-063d2a7435a4:1
streamUrl: https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=258196818&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=6fe8ccce6c53dafa52946f2272ffecea11f4d2635b2fd218d95326739cd96c53
streamId: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#36eaf4b1-8df8-4707-89cf-063d2a7435a4:1
playBehavior: REPLACE_PREVIOUS
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Sending Playback Progress Report Request...
Player State: State.Ended
Sending Playback Progress Report Request...
Play_Audio Request for: https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=258196818&c=cf&f=ts&t=10&bl=256k&s=true&e1=1458835200000&e2=1458836100000&v=V2&h=6fe8ccce6c53dafa52946f2272ffecea11f4d2635b2fd218d95326739cd96c53
Playback Progress Report was Successful!
 Player State: State.Opening

**Comment by lennysh** [comment]
Does running it with sudo help?

**Comment by lennysh** [comment]
Did you run the setup script?  Or did you manually install the new requirements?

**Comment by dliloch** [comment]
I ran the setup script .. sudo ./setup.sh and then rebooted ..

**Comment by lennysh** [comment]
I'm assuming then it also installed the service? Did you stop the service?

**Comment by dliloch** [comment]
I could run the script again and capture the output if that would help...

**Comment by lennysh** [comment]
You may have two Alexa services running now all trying to do the same thing

**Comment by maso27** [comment]
I just updated, and so far it's working great!
She is a bit slower, but right now I'm grooving to some random jazz. I'm waiting for a second song to come around.

Here's how I did the update (not a reinstall):
1.) killall python, to stop the script
2.) place the new requirements.txt, setup.sh, and main.py in the AlexaPi folder
3.) apt-get update, apt-get upgrade, apt-get install vlc
4.) copy and paste the "wget..." line from the top of setup.sh
5.) "pip install-r requirements.txt" from setup.sh
6.) reboot

Great work!

**Comment by lennysh** [comment]
Is the audio device listed differently on the RasPi3?  I've yet to purchase one but wouldn't of thought that it would have been different...  Might be why @dliloch is having an issue with audio playback...

**Comment by dliloch** [comment]
cleared out creds.py .. reinstalled , rebooted ... and bingo .. I'm good ..
response is fine now... thanks amazing ..
:-)
wow

**Comment by lennysh** [comment]
Glad to hear it @dliloch !

**Comment by dliloch** [comment]
the output is coming out of my HDMI .. how can I redirect to the USB audio
dongle which goes to my Bose audio ...
The original main.py went out through the usb audio ...

**Comment by lennysh** [comment]
The following line is the one used to determine the audio output device.  We need to tweak that line (second set of quotes) for your needs.

`i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')`

**Comment by lennysh** [comment]
@dliloch What's the output of the below command?

`aplay -l`

**Comment by maso27** [comment]
Yeah I can't help on audio.  I started from an install of Pi MusicBox so I could get Mopidy and AirPlay support as well.  You can configure the audio output from within the MusicBox webGUI.
But if you've got a desktop installed, can't you just open the ALSA mixer and set it from there?

So I take back what I said about her being slower.  I'd say her response time is comparable.

And this may be feedback for a new "Issue", but I can't pause or stop Audible books after they've started.  They do play, though.

This is really stellar work, @lennysh.

**Comment by dliloch** [comment]
dmux errors ? is this important .. again it was because I am using
debug=1..see attached
by the way .. If I use the word album .. it will play an entire album ..
ie. alexa play stevie nicks album ..
playing popular stevie nicks album! ...

On Thu, Mar 24, 2016 at 1:09 PM, Lenny Shirley <notifications@github.com>
wrote:

> Glad to hear it @dliloch <https://github.com/dliloch> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/sammachin/AlexaPi/issues/20#issuecomment-200930210>
>

If ALEXA selects from "music library" get nothing .. must come from prime music..
sound does not have much bass ??
, expected 4) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 4) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 4) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 4) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 5) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 5) for PID 4096
[6adf5620] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 14) for PID 4096
[6e542898] ts demux error: libdvbpsi (PSI decoder): TS discontinuity (received 0, expected 5) for PID 4096

**Comment by lennysh** [comment]
@maso27 Simply pressing the record button and releasing it immediately "should" stop any current playback.  Not worked on a pause function yet, although a new button could be added to do a pause.

@dliloch The demux error's are unstoppable at the moment from any/all Amazon Music streams regardless if debug is on or not. Still researching that one, but seems to be related to only their streams.  For example, playing from iHeartRadio doesn't return those errors.

**Comment by lennysh** [comment]
@dliloch What's the output of the below command?

`aplay -l`

**Comment by maso27** [comment]
@lennysh yes, playback of music definitely stops when the button is pressed.  The audible book does not.  I've reproduced it twice and had to reboot the Pi to recover.

**Comment by lennysh** [comment]
If anyone prefers live chat, I'm currently online [here](https://webchat.freenode.net/?channels=AlexaDotNet).

**Comment by lennysh** [comment]
@maso27 Can you turn debug mode on, start playing the audio book, and then press the button to attempt to stop playback, and then copy/paste the output here?

**Comment by maso27** [comment]
Here it is.  I pushed the button twice, with no command the first time, then said "Stop" the second time. Also, I think I can hear Alexa talk over the book when I say "Stop":

`Sending Playback Progress Report Request...
Play_Audio Request for: response.mp3
 Play_Audio Request for: https://audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=29477&end=34349&hdnea=st=1458841649~exp=1458841709~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=29477,34349~hmac=6129fedd63cd2f8c8ef26e53da697259be55cf33bf267ea0b396fa44503a887f
[0x74f03c78] inhibit interface error: Failed to connect to the D-Bus session daemon: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11
[0x74f03c78] main interface error: no suitable interface module
[0x75b100e8] inhibit interface error: Failed to connect to the D-Bus session daemon: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11
[0x75b100e8] main interface error: no suitable interface module
Player State: State.Opening
Player State: State.Opening
Player State: State.Playing
Playback Progress Report was Successful!
Player State: State.Ended
[0x2245f30] stream_filter_httplive stream: HTTP Live Streaming (audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=29477&end=34349&hdnea=st=1458841649~exp=1458841709~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=29477,34349~hmac=6129fedd63cd2f8c8ef26e53da697259be55cf33bf267ea0b396fa44503a887f)
[0x2245f30] stream_filter_httplive stream: Meta playlist
[0x2245f30] stream_filter_httplive stream: bandwidth adaptation detected (program-id=1, bandwidth=30000).
Sending GetNextItem Request...
[0x2245f30] stream_filter_httplive stream: VOD Playlist HLS protocol version: 3
[0x2245f30] stream_filter_httplive stream: video on demand (vod) mode
[0x2245f30] stream_filter_httplive stream: Choose segment 0/488 (sequence=2951)
[0x2245f30] stream_filter_httplive stream: downloaded segment 2948 from stream 0
[0x2245f30] stream_filter_httplive stream: downloaded segment 2949 from stream 0
[0x2245f30] stream_filter_httplive stream: playing segment 2948 from stream 0
libdvbpsi error (PSI decoder): TS discontinuity (received 4, expected 0) for PID 0
[0x247b640] ts demux error: MPEG-4 descriptor not found
[0x2245f30] stream_filter_httplive stream: downloaded segment 2950 from stream 0
Player State: State.Playing
[0x2477c10] packetizer_mpeg4audio packetizer: AAC channels: 1 samplerate: 22050
[0x2245f30] stream_filter_httplive stream: downloaded segment 2951 from stream 0
[0x2245f30] stream_filter_httplive stream: downloaded segment 2952 from stream 0
[0x2245f30] stream_filter_httplive stream: downloaded segment 2953 from stream 0
[0x2245f30] stream_filter_httplive stream: downloaded segment 2954 from stream 0
[0x2245f30] stream_filter_httplive stream: playing segment 2949 from stream 0
libdvbpsi error (PSI decoder): TS duplicate (received 0, expected 1) for PID 4095
[0x2245f30] stream_filter_httplive stream: downloaded segment 2955 from stream 0
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"audioItem":{"streams":[{"streamUrl":"https://audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=34349&end=38998&hdnea=st=1458841660~exp=1458841720~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=34349,38998~hmac=5b83763c5b90ffb87d5f99ec3ea3a9a80a8b57acee2bdd51d9efdb42652fe715","offsetInMilliseconds":0,"expiryTime":"2016-03-24T17:48:40+0000","progressReport":{"progressReportDelayInMilliseconds":0,"progressReportIntervalInMilliseconds":30000},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.AudibleClientId#ACRI#url#ACRI#dc54cb67-2c7c-40a3-9d74-a44df3ace1b5:9"}],"audioItemId":"amzn1.as-ct.v1.AudibleClientId#ACRI#url#ACRI#dc54cb67-2c7c-40a3-9d74-a44df3ace1b5:9"},"navigationToken":"amzn1.as-ct.v1.AudibleClientId#ACRI#url#ACRI#dc54cb67-2c7c-40a3-9d74-a44df3ace1b5:9"}}
navigationToken: amzn1.as-ct.v1.AudibleClientId#ACRI#url#ACRI#dc54cb67-2c7c-40a3-9d74-a44df3ace1b5:9
streamUrl: https://audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=34349&end=38998&hdnea=st=1458841660~exp=1458841720~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=34349,38998~hmac=5b83763c5b90ffb87d5f99ec3ea3a9a80a8b57acee2bdd51d9efdb42652fe715
streamId: amzn1.as-ct.v1.AudibleClientId#ACRI#url#ACRI#dc54cb67-2c7c-40a3-9d74-a44df3ace1b5:9
Play_Audio Request for: https://audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=34349&end=38998&hdnea=st=1458841660~exp=1458841720~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=34349,38998~hmac=5b83763c5b90ffb87d5f99ec3ea3a9a80a8b57acee2bdd51d9efdb42652fe715
[0x2527fc8] inhibit interface error: Failed to connect to the D-Bus session daemon: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11
[0x2527fc8] main interface error: no suitable interface module
Recording...
Player State: State.Opening
[0x2245f30] stream_filter_httplive stream: playing segment 2950 from stream 0
libdvbpsi error (PSI decoder): TS duplicate (received 0, expected 1) for PID 4095
[0x2245f30] stream_filter_httplive stream: downloaded segment 2956 from stream 0
[0x2528ac0] stream_filter_httplive stream: HTTP Live Streaming (audiblecdn-vh.akamaihd.net/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/master.m3u8?start=34349&end=38998&hdnea=st=1458841660~exp=1458841720~acl=/i/295890/audiblewords/content/bk/adbl/023112/V$174611$V/aax/bk_adbl_023112_22_32.mp4/*~data=34349,38998~hmac=5b83763c5b90ffb87d5f99ec3ea3a9a80a8b57acee2bdd51d9efdb42652fe715)
[0x2528ac0] stream_filter_httplive stream: Meta playlist
[0x2528ac0] stream_filter_httplive stream: bandwidth adaptation detected (program-id=1, bandwidth=30000).
Recording Finished.
Sending Speech Request...
[0x2528ac0] stream_filter_httplive stream: VOD Playlist HLS protocol version: 3
[0x2528ac0] stream_filter_httplive stream: video on demand (vod) mode
[0x2528ac0] stream_filter_httplive stream: Choose segment 0/466 (sequence=3438)
Processing Request Response...
Request Response is null (This is OKAY!)
[0x2528ac0] stream_filter_httplive stream: downloaded segment 3435 from stream 0
[0x2528ac0] stream_filter_httplive stream: downloaded segment 3436 from stream 0
[0x2528ac0] stream_filter_httplive stream: playing segment 3435 from stream 0
libdvbpsi error (PSI decoder): TS discontinuity (received 11, expected 0) for PID 0
[0x26d7670] ts demux error: MPEG-4 descriptor not found
Player State: State.Playing
[0x2528ac0] stream_filter_httplive stream: playing segment 3436 from stream 0
libdvbpsi error (PSI decoder): TS duplicate (received 0, expected 1) for PID 4095
[0x24fe590] packetizer_mpeg4audio packetizer: AAC channels: 1 samplerate: 22050
Recording...
[0x2528ac0] stream_filter_httplive stream: downloaded segment 3437 from stream 0
Player State: State.Stopped
Recording Finished.
Sending Speech Request...
Processing Request Response...
Request Response is null (This is OKAY!)
[0x2245f30] stream_filter_httplive stream: playing segment 2951 from stream 0
libdvbpsi error (PSI decoder): TS duplicate (received 0, expected 1) for PID 4095
[0x2245f30] stream_filter_httplive stream: downloaded segment 2957 from stream 0
`

**Comment by lennysh** [comment]
@maso27  Hmm...  Audible streaming might be a little different than music...  Just enough to cause problems.  I may have to get a trial period or something in order to do some testing myself.

**Comment by dliloch** [comment]
*pi@raspialexa*:*~ $* aplay -l

**** List of PLAYBACK Hardware Devices ****

card 0: ALSA [bcm2835 ALSA], device 0: bcm2835 ALSA [bcm2835 ALSA]

  Subdevices: 7/8

  Subdevice #0: subdevice #0

  Subdevice #1: subdevice #1

  Subdevice #2: subdevice #2

  Subdevice #3: subdevice #3

  Subdevice #4: subdevice #4

  Subdevice #5: subdevice #5

  Subdevice #6: subdevice #6

  Subdevice #7: subdevice #7

card 0: ALSA [bcm2835 ALSA], device 1: bcm2835 ALSA [bcm2835 IEC958/HDMI]

  Subdevices: 1/1

  Subdevice #0: subdevice #0

card 1: Device [Generic USB Audio Device], device 0: USB Audio [USB Audio]

  Subdevices: 1/1

  Subdevice #0: subdevice #0

**Comment by dliloch** [comment]
if you ask for news it will start but does not get past the weather ... it
is some sort of audio stream ..

**Comment by lennysh** [comment]
@dliloch change
`i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')`
To
`i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=Device,DEV=0')`

**Comment by lennysh** [comment]
@dliloch yeah, news was working the other day, but recently stopped. Still looking into that one.

**Comment by dliloch** [comment]
Lenny,,
perfect! ALSA for HDMI and Device for USB !  ... great work!
-don

**Comment by lennysh** [comment]
@dliloch I just merged in pull request #7 into my [music-patch-1](https://github.com/lennysh/AlexaPi/tree/lennysh-music-patch-1) branch.  Please let me know how it performs in comparison.

**Comment by lennysh** [comment]
All, I'm still available for live chat [here](https://webchat.freenode.net/?channels=AlexaPi).

**Comment by dliloch** [comment]
Lenny,
Seems faster .. I had to simply change the ALSA to Device .. you might make
it a start variable .. and default to ALSA .. no big deal though.
Asked about the new and ALEXA said getting NPR radio .. but I just got the
weather in Seattle.. It would be nice to have default location to where I
am actually at..
still get the demux error ..
when I play a album there seems to be a long wait between songs ..tracks
..but just played an album and you seem to have fixed that..
Again .. great work .. I'm going to put this in a case and replace the
radio on our kitchen sink! Thanks so much for you efforts .. If
I can provide any further help let me know..
-don

**Comment by lennysh** [comment]
@dliloch For the location fix, you need to add your Address under your raspberry pi Alexa device on the Alexa app itself.

**Comment by dliloch** [comment]
Lenny,
added my address and time, weather worked great thanks..

**Comment by MiniguyBrendan** [comment]
Hey, what information did you put for the "buffering/caching mechanisms" on the developer amazon website?

**Comment by lennysh** [comment]
Just VLC Media Player

**Comment by dliloch** [comment]
works  great .. just set it up from crontab @reboot  python main.py ...
running headless .. when I hear the hello I am ready to go ...
still running on a pi3 which is a waste of resources .. going to try and
run on an old banana pi with integrated mic and wifi.. that would
make a perfect small form factor box..
-don

**Comment by MiniguyBrendan** [comment]
They already denied it and there's no edit button to add VLC Media Player :( any ideas?

**Comment by lennysh** [comment]
Just change the text and resave

**Comment by MiniguyBrendan** [comment]
It's grayed out though.

**Comment by MiniguyBrendan** [comment]
I think it's because it says it's pending. But amazon already sent me an email saying it was rejected. Weird.

**Comment by lennysh** [comment]
Send me a screenshot

**Comment by MiniguyBrendan** [comment]
http://imgur.com/UwLl01P

I didn't know it was required so I initially had just put "Doesn't have." I also didn't know what it meant, let alone that VLC handles that.

**Comment by lennysh** [comment]
And you can't type new text in the boxes?

**Comment by MiniguyBrendan** [comment]
Nope

**Comment by MiniguyBrendan** [comment]
Should I just create a new Alexa product?

**Comment by lennysh** [comment]
No, just wait...  They may have to update the listing as rejected in order to make changes.  I got rejected the first two times I submitted it, lol.

**Comment by SpiderSpartan** [comment]
Thanks for getting this working.  I'm new to git and github, so I had to download your branch as a zip to get all the files working, but it does work.  One thing I noticed is the newsflash freaks out, playing stories on top of each other.  The nova spirit fork (I think) I was using before just stopped after saying NPR.  I'm loving the music though.  Thanks again.

**Comment by lennysh** [comment]
I'll try to fix the flash briefing issue tomorrow.

**Comment by maso27** [comment]
As to the Audible ebooks: it seems to be playing and pausing now.
I'm not sure what the cause was, but here's what changed on my side of things:

I installed the Audible app on my Android phone and logged in.
Previously I had been using the "Whispersync" functionality within the Kindle app, so nothing was officially connected to Audible.

I'm not sure if that changed anything, or if Amazon changed something behind the scenes, or what.  But it works!

**Comment by ggavalas** [comment]
I posted a partial fix that worked for me on the flash briefings playing on top of each other over in issue 31, here: [https://github.com/sammachin/AlexaPi/issues/31](url).  It doesn't make tune-in stations work, but the news read by Alexa does not play on top of each other anymore.

**Comment by dliloch** [comment]
Thanks for the update ..it works great... Plays for about 5 minutes finishing with the weather...

**Comment by lennysh** [comment]
Anybody feel like trying a code change for me?  Don't have my Pi with me atm...  Possible TuneIn fix...

**Comment by dliloch** [comment]
Lenny,
I'm ready to test..let me know what to do...

**Comment by lennysh** [comment]
1. Turn on debug.
2. Replace def play_audio(file) with the following code:

```
def play_audio(file):
	global nav_token, p, audioplaying
	if debug: print("{}Play_Audio Request for:{} {}".format(bcolors.OKBLUE, bcolors.ENDC, file))
	GPIO.output(24, GPIO.HIGH)
	ext = (file.rpartition(".")[2])[:3]
	#subprocess.Popen(['mpg123', '-q', '{}{}'.format(path, file)]).wait()	
	i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')
	mrl = ""
	if file == "response.mp3" or file == "hello.mp3":
		mrl = "{}{}".format(path, file)
	else:
		mrl = "{}".format(file)
	
	if mrl != "":
		m = i.media_new(mrl)
		ml = i.media_list_new(mrl)
		m.get_mrl()
		mm = m.event_manager()
		#mm.event_attach(vlc.EventType.MediaPlayerTimeChanged, pos_callback)
		#mm.event_attach(vlc.EventType.MediaParsedChanged, meta_callback, m)
		audioplaying = True
		if ext in playlists:
			p = i.media_list_player_new()
			p.set_media_list(ml)
			if debug: print("{}Requesting Playlist...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		else:
			p = i.media_player_new()
			p.set_media(m)
			if debug: print("{}Requesting Stream...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		mm.event_attach(vlc.EventType.MediaStateChanged, state_callback, p)
		p.audio_set_volume(100)
		p.play()
		while audioplaying:
			continue
		GPIO.output(24, GPIO.LOW)
	else:
		print("(play_audio) mrl = Nothing!")
```

**Comment by dliloch** [comment]
Get playlists not found?

**Comment by lennysh** [comment]
Try this one:
```
def play_audio(file):
	global nav_token, p, audioplaying
	if debug: print("{}Play_Audio Request for:{} {}".format(bcolors.OKBLUE, bcolors.ENDC, file))
	GPIO.output(24, GPIO.HIGH)
	ext = (file.rpartition(".")[2])[:3]
	#subprocess.Popen(['mpg123', '-q', '{}{}'.format(path, file)]).wait()	
	i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')
	mrl = ""
	if file == "response.mp3" or file == "hello.mp3":
		mrl = "{}{}".format(path, file)
	else:
		mrl = "{}".format(file)
	
	if mrl != "":
		if ext in playlists:
			#m.get_mrl()
			m = i.media_list_new(mrl)
			p = i.media_list_player_new()
			p.set_media_list(ml)
			if debug: print("{}Requesting Playlist...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		else:
			m = i.media_new(mrl)
			p = i.media_player_new()
			p.set_media(m)
			if debug: print("{}Requesting Stream...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		audioplaying = True
		mm = m.event_manager()
		#mm.event_attach(vlc.EventType.MediaPlayerTimeChanged, pos_callback)
		#mm.event_attach(vlc.EventType.MediaParsedChanged, meta_callback, m)
		mm.event_attach(vlc.EventType.MediaStateChanged, state_callback, p)
		p.audio_set_volume(100)
		p.play()
		while audioplaying:
			continue
		GPIO.output(24, GPIO.LOW)
	else:
		print("(play_audio) mrl = Nothing!")
```

**Comment by dliloch** [comment]
pi@raspberrypi:~/AlexaPiP1 $ python main2.py
Checking Internet Connection...
Connection OK
Play_Audio Request for: hello.mp3
Traceback (most recent call last):
 File "main2.py", line 416, in <module>
   play_audio("hello.mp3")
 File "main2.py", line 246, in play_audio
   if ext in playlists:
NameError: global name 'playlists' is not defined

**Comment by lennysh** [comment]
My bad @dliloch ...  Add the following line to the top of the file under the "Settings" section...
`playlists = set(['pls','m3u'])`

**Comment by dliloch** [comment]
Ok..that worked for play headline news..but I lost the hello audio message when I started it and when I asked Alexa to play an album I got this.. The python script did not crash but no playback..
Play_Audio Request for:
https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460574900000&e2=1460575800000&v=V2&h=b048f403a3acbfeaa0fc5f5ee175e81c9f3c85e09c051b7043119809ab5ad08f
Exception in thread Thread-24:
Traceback (most recent call last):
 File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
   self.run()
 File "/usr/lib/python2.7/threading.py", line 763, in run
   self.__target(*self.__args, **self.__kwargs)
 File "main2.py", line 251, in play_audio
   p.set_media_list(ml)
NameError: global name 'ml' is not defined

**Comment by lennysh** [comment]
Ugh, losing my touch of offline code changes...

Change
`p.set_media_list(ml)`
to
`p.set_media_list(m)`

**Comment by dliloch** [comment]
Still no hello audio 
And for play of album..
Ready to Record.
Sending Playback Progress Report Request...
Play_Audio Request for:
https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460575800000&e2=1460576700000&v=V2&h=c9bcbc5503f712d1788e3b3f03681700969dc88ed4e889b08d20f09ef6b0bfeb
Requesting Playlist...
This object event manager doesn't know about 'MediaStateChanged'
eventsException in thread Thread-4:
Traceback (most recent call last):
 File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
   self.run()
 File "/usr/lib/python2.7/threading.py", line 763, in run
   self.__target(*self.__args, **self.__kwargs)
 File "main2.py", line 263, in play_audio
   p.audio_set_volume(100)
AttributeError: 'MediaListPlayer' object has no attribute 'audio_set_volume'

Playback Progress Report was Successful!

**Comment by dliloch** [comment]
Got the hello to work... Lost the CARD=Device..so that is ok

**Comment by lennysh** [comment]
If the VLC List Player doesn't have a MediaStateChanged event, then we can't use it (don't think)...  Ugh...

The volume part can be alleviated (for the moment) by moving that line into the else: section of the ext in playlists if statement.

**Comment by lennysh** [comment]
For example, try this one:
```
def play_audio(file):
	global nav_token, p, audioplaying
	if debug: print("{}Play_Audio Request for:{} {}".format(bcolors.OKBLUE, bcolors.ENDC, file))
	GPIO.output(24, GPIO.HIGH)
	ext = (file.rpartition(".")[2])[:3]
	#subprocess.Popen(['mpg123', '-q', '{}{}'.format(path, file)]).wait()	
	i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')
	mrl = ""
	if file == "response.mp3" or file == "hello.mp3":
		mrl = "{}{}".format(path, file)
	else:
		mrl = "{}".format(file)
	
	if mrl != "":
		if ext in playlists:
			#m.get_mrl()
			m = i.media_list_new(mrl)
			p = i.media_list_player_new()
			p.set_media_list(m)
			if debug: print("{}Requesting Playlist...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		else:
			m = i.media_new(mrl)
			p = i.media_player_new()
			p.set_media(m)
			mm = m.event_manager()
			#mm.event_attach(vlc.EventType.MediaPlayerTimeChanged, pos_callback)
			#mm.event_attach(vlc.EventType.MediaParsedChanged, meta_callback, m)
			mm.event_attach(vlc.EventType.MediaStateChanged, state_callback, p)
			p.audio_set_volume(100)
			if debug: print("{}Requesting Stream...{}".format(bcolors.OKBLUE, bcolors.ENDC))
		audioplaying = True
		p.play()
		while audioplaying:
			continue
		GPIO.output(24, GPIO.LOW)
	else:
		print("(play_audio) mrl = Nothing!")
```

**Comment by lennysh** [comment]
Might work...  Try TuneIn and see how it does for playing past the first song...  Hoping it's a continuous stream instead of different playlist per song...

**Comment by dliloch** [comment]
No luck here ..different errors..
Ready to Record.
Sending Playback Progress Report Request...
Play_Audio Request for:
https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460577600000&e2=1460578500000&v=V2&h=efd2d2ace6b5c9f64d9808053a6d7cf8fd04c69dae806be5c0ebeaf41794cf2d
Requesting Playlist...
[7064a9a0] pulse audio output error: PulseAudio server connection
failure: Connection refused
[74517b98] filesystem access error: cannot open file
/home/pi/AlexaPiP1/h (No such file or directory)
[74517b98] core access error: File reading failed
[74517b98] core access error: VLC could not open the file
"/home/pi/AlexaPiP1/h" (No such file or directory).
[7066d9b8] core input error: open of `file:///home/pi/AlexaPiP1/h' failed
[7066d9b8] core input error: Your input can't be opened
[7066d9b8] core input error: VLC is unable to open the MRL
'file:///home/pi/AlexaPiP1/h'. Check the log for details.
Playback Progress Report was Successful!

**Comment by lennysh** [comment]
Does other audio stuff still work? iHeartRadio, Amazon, etc...

**Comment by dliloch** [comment]
Headline works..Iheart works but get these errors..
Sending Playback Progress Report Request...
[6f7d7428] alsa audio output error: cannot open ALSA device
"hw:CARD=Device,DEV=0": Device or resource busy
[6f7d7428] core audio output error: Audio output failed
[6f7d7428] core audio output error: The audio device
"hw:CARD=Device,DEV=0" could not be used:
Device or resource busy.
[6f7d7428] core audio output error: module not functional
[7071b1b0] core decoder error: failed to create audio output
Playback Progress Report was Successful!

**Comment by dliloch** [comment]
Tunein does not work seems to loop

**Comment by lennysh** [comment]
Here's the one I'm currently using:
```
def play_audio(file):
	global nav_token, p, audioplaying
	if debug: print("{}Play_Audio Request for:{} {}".format(bcolors.OKBLUE, bcolors.ENDC, file))
	GPIO.output(24, GPIO.HIGH)
	ext = (file.rpartition(".")[2])[:3]
	if debug: print("{}Variable 'ext' ={} {}".format(bcolors.OKBLUE, bcolors.ENDC, ext))
	#subprocess.Popen(['mpg123', '-q', '{}{}'.format(path, file)]).wait()	
	i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')
	mrl = ""
	type = ""
	if ext == "mp3":
		mrl = "{}{}".format(path, file)
		type = "File"
	else:
		mrl = "{}".format(file)
		type = "Stream"
	
	if debug: print("{}Variable 'mrl' ={} {}".format(bcolors.OKBLUE, bcolors.ENDC, mrl))
	
	if mrl != "":
		if ext in playlists:
			type = "Playlist"
			m = i.media_list_new([file])
			#m.get_mrl()
			p = i.media_list_player_new()
			p.set_media_list(m)
		else:
			m = i.media_new(mrl)
			p = i.media_player_new()
			p.set_media(m)
			mm = m.event_manager()
			#mm.event_attach(vlc.EventType.MediaPlayerTimeChanged, pos_callback)
			#mm.event_attach(vlc.EventType.MediaParsedChanged, meta_callback, m)
			mm.event_attach(vlc.EventType.MediaStateChanged, state_callback, p)
			p.audio_set_volume(100)
		audioplaying = True
		if debug: print("{}Requesting {}...{}".format(bcolors.OKBLUE, type, bcolors.ENDC))
		p.play()
		while audioplaying:
			continue
		GPIO.output(24, GPIO.LOW)
	else:
		print("(play_audio) mrl = Nothing!")
```
but now it's showing this as a result
```
Checking Internet Connection...
Connection OK
Play_Audio Request for: hello.mp3
Variable 'ext' = mp3
Variable 'mrl' = /usr/share/AlexaPi/hello.mp3
Requesting File...
Player State: State.Opening
Player State: State.Playing
Player State: State.Ended
Play_Audio Request for: https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460577600000&e2=1460578500000&v=V2&h=efd2d2ace6b5c9f64d9808053a6d7cf8fd04c69dae806be5c0ebeaf41794cf2d
Variable 'ext' = m3u
Variable 'mrl' = https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460577600000&e2=1460578500000&v=V2&h=efd2d2ace6b5c9f64d9808053a6d7cf8fd04c69dae806be5c0ebeaf41794cf2d
Requesting Playlist...
[016aa838] pulse audio output error: PulseAudio server connection failure: Connection refused
[76003388] http access error: error: HTTP/1.1 403 Forbidden
[76003388] http access error: error: HTTP/1.1 403 Forbidden
[01671950] core input error: open of `https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460577600000&e2=1460578500000&v=V2&h=efd2d2ace6b5c9f64d9808053a6d7cf8fd04c69dae806be5c0ebeaf41794cf2d' failed
[01671950] core input error: Your input can't be opened
[01671950] core input error: VLC is unable to open the MRL 'https://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m=m&dmid=233410175&c=cf&f=ts&t=10&bl=256k&s=true&e1=1460577600000&e2=1460578500000&v=V2&h=efd2d2ace6b5c9f64d9808053a6d7cf8fd04c69dae806be5c0ebeaf41794cf2d'. Check the log for details.
```
Getting a 403 Forbidden message back...  Hmm...

**Comment by ggavalas** [comment]
I'm not able to help, ATM, but I did notice when I was attempting to work with the tunein streams the other day that there was a partnerid attached to the end of the streamurl.  Oddly it had an exclamation point at the beginning of the id.  I'm not seeing that attached to your m3u url strings, however.  I wonder if this might be part of the forbidden issue.  Just a thought, and possible useless remark.    

```
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_dcf6a4a3-bf88-4625-b474-d531cc1ce700","audioContent":"cid:DeviceTTSRendererV4_dcf6a4a3-bf88-4625-b474-d531cc1ce700_749573739"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"https://opml.radiotime.com/Tune.ashx?serial=AFTY7K6EBGRNINQQDXYGC5WUC7AQ&formats=aac,mp3&partnerId=!EALLOjB&id=s21392","offsetInMilliseconds":0,"expiryTime":"2016-04-13T20:18:57+0000","progressReport":{"progressReportDelayInMilliseconds":0,"progressReportIntervalInMilliseconds":900000},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#c2b4eb3b-a5df-444a-a46b-1ba438af188d:1"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#c2b4eb3b-a5df-444a-a46b-1ba438af188d:1"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#c2b4eb3b-a5df-444a-a46b-1ba438af188d:1","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#c2b4eb3b-a5df-444a-a46b-1ba438af188d:1
streamUrl: https://opml.radiotime.com/Tune.ashx?serial=AFTY7K6EBGRNINQQDXYGC5WUC7AQ&formats=aac,mp3&partnerId=!EALLOjB&id=s21392
```

**Comment by ggavalas** [comment]
it seems as though it was reading through the m3u file AS the media and ending without playing on some of the streams.  I'm still not sure about the 403 forbidden issue.  I did get radio station streams and podcasts to work (but not npr, bbc, etc from flash briefings) with the following changes for me.  Basically, it is flagged when it reaches the end of the m3u file then creates a sublist of items to play from the m3u itself.  The code was modified from a forum response [here](https://forum.videolan.org/viewtopic.php?t=109348#p370825).  

I added an ash (really for ashx) extension to the playlists line in the settings at the top of the code:

```
playlists = set(['pls','m3u','ash']) 
```

I also added an end_reached definition to notify when the m3u is finished and added a global variable for flag then made the following changes to the play_audio definition:

```
#Added this to Global project
flag = 0

def end_reached(self):
    global flag
    flag = 1
    print("End reached!")
#End


def play_audio(file):
    global nav_token, p, audioplaying, flag #<-added flag here
    if debug: print("{}Play_Audio Request for:{} {}".format(bcolors.OKBLUE, bcolors.ENDC, file))
    GPIO.output(24, GPIO.HIGH)
    ext = (file.rpartition(".")[2])[:3]
    #subprocess.Popen(['mpg123', '-q', '{}{}'.format(path, file)]).wait()
    i = vlc.Instance('--aout=alsa', '--alsa-audio-device=hw:CARD=ALSA,DEV=0')
    mrl = ""
    print(ext)
    if file == "response.mp3" or file == "hello.mp3":
        mrl = "{}{}".format(path, file)
    else:
        mrl = "{}".format(file)

    if mrl != "":
        if ext in playlists:

            #Replaced code here
            p=i.media_player_new() # Create new media player
            event_manager = p.event_manager() # Attach event to player (next 3 lines)
            event=vlc.EventType()
            event_manager.event_attach(event.MediaPlayerEndReached, end_reached)
            m=i.media_new(mrl) # Create new media
            p.set_media(m) # Set URL as the player's media
            m.release()
            p.play() # play it
            while flag == 0: # Wait until the end of the first media has been reached.$
                time.sleep(0.5)
                if debug: print("{}Loading Playlist...{}".format(bcolors.OKBLUE, bcolors.ENDC))
            sub_list = m.subitems() # .. and get the sub itmes in the playlist
            sub_list.lock()
            sub = sub_list.item_at_index(0) # Get the first sub item
            sub_list.unlock()
            sub_list.release()
            p.set_media(sub) # Set it as the new media in the player
            #End of replaced Code
        else:
            m = i.media_new(mrl)
            p = i.media_player_new()
            p.set_media(m)
            mm = m.event_manager()
            #mm.event_attach(vlc.EventType.MediaPlayerTimeChanged, pos_callback)
            #mm.event_attach(vlc.EventType.MediaParsedChanged, meta_callback, m)
            mm.event_attach(vlc.EventType.MediaStateChanged, state_callback, p)
            p.audio_set_volume(100)
            if debug: print("{}Requesting Stream...{}".format(bcolors.OKBLUE, bcolors.ENDC))
        audioplaying = True
        p.play()
        while audioplaying:
            continue
        GPIO.output(24, GPIO.LOW)
    else:
        print("(play_audio) mrl = Nothing!")
```

At least it is a start to resolving this issue.  Like I said, the TuneIn streams seem to be working, now, but the flash briefing isn't.

**Comment by ggavalas** [comment]
lol, sorry about leaving my own debug prints in there for the extensions and the end of m3u reading.

**Comment by dony71** [comment]
lennysh,
i try to play iheart radio, it starts like 1/2 second then stop
can you help to check what's wrong?
below is debug dump

Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080","audioContent":"cid:DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080_1324119209"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us","offsetInMilliseconds":0,"expiryTime":"2016-04-26T01:58:04+0000","progressReport":{"progressReportDelayInMilliseconds":15000,"progressReportIntervalInMilliseconds":60000},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
streamUrl: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
streamId: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
playBehavior: REPLACE_PREVIOUS
Play_Audio Request for: response.mp3
Player State: State.OpeningReady to Record.
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080","audioContent":"cid:DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080_1324119209"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us","offsetInMilliseconds":0,"expiryTime":"2016-04-26T01:58:04+0000","progressReport":{"progressReportDelayInMilliseconds":15000,"progressReportIntervalInMilliseconds":60000},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
streamUrl: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
streamId: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
playBehavior: REPLACE_PREVIOUS
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Sending Playback Progress Report Request...
Playback Progress Report was Successful!
Player State: State.Ended
Sending Playback Progress Report Request...
Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
 Play_Audio Request for: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
Player State: State.Opening
Player State: State.Playing
Playback Progress Report was Successful!
Processing Request Response...
Request Response is null (This is OKAY!)
Ready to Record.
Player State: State.Stopped

Player State: State.Playing
Sending Playback Progress Report Request...
Playback Progress Report was Successful!
Player State: State.Ended
Sending Playback Progress Report Request...
Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
 Play_Audio Request for: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
Processing Request Response...
JSON String Returned: {"messageHeader":{},"messageBody":{"directives":[{"namespace":"SpeechSynthesizer","name":"speak","payload":{"contentIdentifier":"amzn1.as-ct.v1.Domain:Application:Music#ACRI#DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080","audioContent":"cid:DeviceTTSRendererV4_5156f955-b784-4b38-bd3d-19aa90b0c080_1324119209"}},{"namespace":"AudioPlayer","name":"play","payload":{"audioItem":{"streams":[{"streamUrl":"http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us","offsetInMilliseconds":0,"expiryTime":"2016-04-26T01:58:04+0000","progressReport":{"progressReportDelayInMilliseconds":15000,"progressReportIntervalInMilliseconds":60000},"progressReportRequired":true,"streamId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"}],"audioItemId":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1"},"navigationToken":"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1","playBehavior":"REPLACE_PREVIOUS"}}]}}
navigationToken: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
streamUrl: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
streamId: amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#43ce9c6f-696d-4688-97b1-d94e0ec82640:1
playBehavior: REPLACE_PREVIOUS
Play_Audio Request for: response.mp3
Player State: State.Opening
Player State: State.Playing
Sending Playback Progress Report Request...
Playback Progress Report was Successful!
Player State: State.Ended
Sending Playback Progress Report Request...
Ready to Record.
Recording...
Recording Finished.
Sending Speech Request...
 Play_Audio Request for: http://akstream2.iheart.com/WMG2/Thumb_Content/Full_AAC/WMG/Feb10/022310/20100223_0210_10/881410001760/881410001760_00008_256.m4a?nga=20160426015804&track=5188539&player=WFEwa1NVMlAwbVRQSFdnd2NSOU1LQT09&env=live&sig=2R-4J13zcYrPIiM2C24-YNx-yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
Player State: State.Opening
Player State: State.Playing
Playback Progress Report was Successful!
Processing Request Response...
Request Response is null (This is OKAY!)
Ready to Record.
Player State: State.Stopped
yjzDsB1DMxl2oqdEF38&as=SIMPLE&rnga=20160426015804&host=amazon.appliance.us
Player State: State.Opening
Player State: State.Playing
Playback Progress Report was Successful!
Processing Request Response...
Request Response is null (This is OKAY!)
Ready to Record.
Player State: State.Stopped

**Comment by jrlambs** [comment]
I feel like I'm missing something.  I've got my the code downloaded and everything running.. debug mode shows no errors.. I hear hello when I start up the script.. but when I go to echo.amazon.com I get the our "Alexa device is not connected" message.  I've downloaded and am using the lennysh-music-patch-1 branch.. and I've turned on/filled out the amazon music section of the amazon voice service device setup. Is it supposed to show as online in echo.alexa.com? how do you set up music?

**Comment by dony71** [comment]
ggavalals, I follow your "def play_audio(file)" code, but got this error below
anything I'm missing?

  File "main.py", line 299
    if mrl != "":
                ^
IndentationError: unindent does not match any outer indentation level

**Comment by ggavalas** [comment]
Sorry I haven't worked more on this lately.  Finishing nursing school at the moment and doing exams. Hopefully I'll be back on it soon. 

That IndentationError is exactly what it sounds like.  
```
if mrl !="": 
```
should be indented the same number of spaces as the 
```
else:
```
That is above it.  It looks like you may have lost some formatting when you copied it all over.

**Comment by dony71** [comment]
ggavalas, thank's
i found my typo mistake

**Comment by sammachin** [comment]
@jrlambs yeah the remote control of music only works for devices using the new v2016 api which has some push capability, I'm working on using that for v2 of AlexaPi so hopefully in the future ;)

@dony71 Yep you'll need to enable Amazon Music on the developer account, it usually takes them about 24hrs to approve you

**Comment by sammachin** [comment]
@lennysh This seems to be a thread mostly around your music patch? 

Are you happy if I close it now as its getting a bit long and hopefully people will open new issues for separate items. I'm going to try and manage the issues better going forward :)

**Comment by lennysh** [comment]
@sammachin Yes, please feel free to close.

**Comment by dliloch** [comment]
Hello,
I waited until the updates seemed stable and then downloaded the new code. It seems to run ok ..ie. Get internet connection OK ..
1) no hello..
2) I can ask for song and Alexa responds but then silence
3) I ask for headline news only one item plays
I get no errors as such..I see debug set to 1 but nothing
If I fall back to the old main.py all is ok..
Any ideas?

**Comment by interbiznw** [comment]
Why do i get "Amazon music is not supported on this device"...? Amazon music is approved on my account..?
